File: etc/dev.py
from typing import Any
import argparse
import os
import re
import shutil
import subprocess
import sys
import time
from pathlib import Path
from subprocess import CalledProcessError

BENCHMARKS = {
    "repos": {
        "k8s": {
            "git": "https://github.com/kubernetes/kubernetes.git",
            "commit": "03e14cc9432975dec161de1e52d7010f9711a913",
        },
        "linux": {
            "git": "https://github.com/torvalds/linux.git",
            "commit": "9d9c1cfec01cdbf24bd9322ed555713a20422115",
        },
        "wikiquote": {
            "git": "https://github.com/federicotdn/wikiquote.git",
            "commit": "2ded998e4f7c5f08e0c66bd34c747a6842a2aedb",
        },
    },
    "findSymbolDefinition": {
        "k8s": {"repo": "k8s", "lang": "go", "symbol": "genMarkdown"},
        "k8s-new": {"repo": "k8s", "lang": "go", "symbol": "New"},
        "k8s-vendor": {"repo": "k8s", "lang": "go", "symbol": "IgnoreTopFunction"},
        "k8s-notfound": {
            "repo": "k8s",
            "lang": "go",
            "symbol": "ThisSymbolDoesNotExist",
        },
        "wq": {"repo": "wikiquote", "lang": "python", "symbol": "quote_of_the_day"},
        "linux": {"repo": "linux", "lang": "c", "symbol": "io_madvise"},
    },
}


def run(*args: Any, verbose: bool = False, **kwargs: Any):
    kwargs.setdefault("check", True)
    if verbose:
        print("dev:", " ".join(args))
    return subprocess.run(args, **kwargs)


def sep() -> None:
    print("-" * 80)


def install(profile: bool = False) -> None:
    install_dir = Path.home() / ".local" / "bin"
    args = [
        "cabal",
        "install",
        "exe:irk",
        "-j",
        f"--installdir={install_dir}",
        "--overwrite-policy=always",
    ]
    if profile:
        args.extend(["--enable-profiling", '--ghc-options="-fprof-late"'])

    try:
        run(*args, verbose=True)
    except CalledProcessError as e:
        if sys.platform == "win32":
            print(
                "hint: the .exe file may currently be running, close it and try again.",
                file=sys.stderr,
            )
        print("error:", e)
        exit(1)


def vendor() -> None:
    run("cabal", "freeze")
    if os.path.exists("vendor"):
        shutil.rmtree("vendor")
    os.makedirs("vendor")

    with open("cabal.project.freeze") as f:
        for line in f:
            if "any." not in line:
                continue

            if match := re.search(r"any\.([^ ]*) ==([^,]*)", line):
                pkg = f"{match.group(1)}-{match.group(2)}"
                run(
                    "cabal",
                    "get",
                    "--destdir=./vendor",
                    pkg,
                    check=False,
                )

    os.remove("cabal.project.freeze")


def benchmark(
    case_name: str,
    profile: bool = False,
    use_rg: bool = False,
) -> None:
    script_dir = Path(__file__).parent
    root_dir = Path.cwd()
    repos_dir = script_dir / "repos"

    if case_name not in BENCHMARKS["findSymbolDefinition"]:
        print("invalid case; available cases:")
        for case in BENCHMARKS["findSymbolDefinition"].keys():
            print(f"  {case}")
        exit(1)

    scenario = BENCHMARKS["findSymbolDefinition"][case_name]
    repo_id = scenario["repo"]
    language = scenario["lang"]
    symbol = scenario["symbol"]

    repo_info = BENCHMARKS["repos"][repo_id]
    git_url = repo_info["git"]
    commit = repo_info["commit"]

    repo_name = Path(git_url).stem
    repo_path = repos_dir / repo_name

    print(f"scenario: {case_name}")
    print(f"  git: {git_url}")
    print(f"  commit: {commit}")
    print(f"  language: {language}")
    print(f"  symbol: {symbol}")
    sep()

    repos_dir.mkdir(exist_ok=True)

    if not repo_path.exists():
        run("git", "clone", "--filter=blob:none", git_url, str(repo_path))

    result = run(
        "git", "rev-parse", "HEAD", cwd=repo_path, capture_output=True, text=True
    )
    current_commit = result.stdout.strip()
    if current_commit != commit:
        run("git", "checkout", commit, cwd=repo_path)

    os.chdir(root_dir)

    if use_rg:
        cmd = [
            "rg",
            "-t",
            language,
            "--word-regexp",
            symbol,
            str(repo_path),
            "--column",
            "--no-heading",
        ]
        print("running rg...")
        run(*cmd)
        sep()
        print("running rg benchmark...")
        cmd_str = " ".join(cmd)
        run("hyperfine", "-i", "--warmup", "2", cmd_str)
    elif profile:
        install(profile=True)

        if os.path.exists("irk.prof"):
            os.remove("irk.prof")

        sep()
        print("profiling irk...")
        cmd = [
            "irk",
            "find",
            "-w",
            str(repo_path),
            "-l",
            language,
            symbol,
            "+RTS",
            "-pj",
            "-RTS",
        ]

        start = time.perf_counter()
        run(*cmd)
        elapsed = time.perf_counter() - start
        print(f"Elapsed time: {elapsed:.3f}s")

        if os.path.exists("irk.prof"):
            timestamp = time.strftime("%Y%m%d-%H%M%S")
            newname = f"irk.{timestamp}.prof"
            os.rename("irk.prof", newname)
            sep()
            print(f"profile saved to {newname}")
    else:
        install()

        sep()
        print("running irk...")
        cmd = ["irk", "find", "-w", str(repo_path), "-l", language, symbol]

        start = time.perf_counter()
        run(*cmd)
        elapsed = time.perf_counter() - start
        print(f"elapsed time: {elapsed:.3f}s")

        sep()
        print("running irk benchmark...")
        cmd_str = " ".join(cmd)
        run("hyperfine", "--warmup", "2", cmd_str)


def main() -> None:
    parser = argparse.ArgumentParser()
    subparsers = parser.add_subparsers(dest="command", required=True)

    subparsers.add_parser("install")
    subparsers.add_parser("vendor")

    bench_parser = subparsers.add_parser("benchmark")
    bench_parser.add_argument("case_name")
    bench_parser.add_argument("--profile", action="store_true")
    bench_parser.add_argument("--rg", action="store_true")

    args = parser.parse_args()

    if args.command == "install":
        install()
    elif args.command == "vendor":
        vendor()
    elif args.command == "benchmark":
        benchmark(args.case_name, args.profile, args.rg)


if __name__ == "__main__":
    main()

File: etc/Dev.hs
module Main where

import Control.Exception (SomeException, try)
import Control.Monad (unless, when)
import Data.List (isInfixOf, isPrefixOf)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (isJust)
import Data.Time.Clock (diffUTCTime, getCurrentTime)
import Data.Time.Format (defaultTimeLocale, formatTime)
import Options.Applicative
import System.Directory
import System.Environment (lookupEnv)
import System.Exit (exitFailure)
import System.FilePath (takeBaseName, (</>))
import System.Info (os)
import System.Process (callCommand)
import Text.Printf (printf)

data RepoInfo = RepoInfo {repoGit :: String, repoCommit :: String}

data Scenario = Scenario
  { scenarioRepo :: String,
    scenarioLang :: String,
    scenarioSymbol :: String
  }

repos :: Map String RepoInfo
repos =
  Map.fromList
    [ ( "k8s",
        RepoInfo
          "https://github.com/kubernetes/kubernetes.git"
          "03e14cc9432975dec161de1e52d7010f9711a913"
      ),
      ( "linux",
        RepoInfo
          "https://github.com/torvalds/linux.git"
          "9d9c1cfec01cdbf24bd9322ed555713a20422115"
      ),
      ( "wikiquote",
        RepoInfo
          "https://github.com/federicotdn/wikiquote.git"
          "2ded998e4f7c5f08e0c66bd34c747a6842a2aedb"
      ),
      ( "ghc",
        RepoInfo
          "https://github.com/ghc/ghc.git"
          "3d6aba778f7628deef0141d9e8effc546c09bac4"
      ),
      ( "rails",
        RepoInfo
          "https://github.com/rails/rails.git"
          "01fb87c25e541ea889263400012bb85e822496f5"
      )
    ]

benchmarks :: Map String Scenario
benchmarks =
  Map.fromList
    [ ("k8s", Scenario "k8s" "go" "genMarkdown"),
      ("k8s-new", Scenario "k8s" "go" "New"),
      ("k8s-vendor", Scenario "k8s" "go" "IgnoreTopFunction"),
      ("k8s-notfound", Scenario "k8s" "go" "ThisSymbolDoesNotExist"),
      ("wq", Scenario "wikiquote" "python" "quote_of_the_day"),
      ("linux", Scenario "linux" "c" "io_madvise"),
      ("ghc", Scenario "ghc" "haskell" "pprWithSourceText"),
      ("rails", Scenario "rails" "ruby" "each_connection_pool")
    ]

data Command = Install | Vendor | Benchmark String Bool Bool

commandParser :: Parser Command
commandParser =
  subparser
    ( command "install" (info (pure Install) (progDesc "Install irk."))
        <> command "vendor" (info (pure Vendor) (progDesc "Vendor dependencies."))
        <> command "benchmark" (info (benchmarkParser <**> helper) (progDesc "Run benchmarks."))
    )

benchmarkParser :: Parser Command
benchmarkParser =
  Benchmark
    <$> argument str (metavar "CASE_NAME")
    <*> switch (long "profile" <> help "Enable profiling.")
    <*> switch (long "rg" <> help "Use ripgrep instead of irk.")

opts :: ParserInfo Command
opts = info (commandParser <**> helper) (fullDesc <> progDesc "Development tools for irk.")

run :: String -> IO ()
run cmd = do
  verbose <- lookupEnv "VERBOSE"
  when (isJust verbose) $ putStrLn $ "running: " ++ cmd
  callCommand cmd

sep :: IO ()
sep = putStrLn $ replicate 80 '-'

install :: Bool -> IO ()
install profile = do
  home <- getHomeDirectory
  let installDir = home </> ".local" </> "bin"
      baseArgs =
        [ "install",
          "exe:irk",
          "-j",
          "--installdir=" ++ installDir,
          "--overwrite-policy=always"
        ]
      profileArgs =
        if profile
          then ["--enable-profiling", "--ghc-options=-fprof-late"]
          else []
      args = baseArgs ++ profileArgs

  result <- try (run $ "cabal " ++ unwords args) :: IO (Either SomeException ())
  case result of
    Left e -> do
      when (os == "mingw32") $
        putStrLn "hint: the .exe file may currently be running, close it and try again."
      putStrLn $ "error: " ++ show e
      exitFailure
    Right () -> return ()

vendor :: IO ()
vendor = do
  run "cabal freeze"

  vendorExists <- doesDirectoryExist "vendor"
  when vendorExists $ removeDirectoryRecursive "vendor"
  createDirectory "vendor"

  contents <- readFile "cabal.project.freeze"
  mapM_ processLine (lines contents)

  removeFile "cabal.project.freeze"
  where
    processLine line
      | "any." `isInfixOf` line = case extractPackage line of
          Just pkg -> run $ "cabal get --destdir=./vendor " ++ pkg ++ " || true"
          Nothing -> return ()
      | otherwise = return ()

    extractPackage line =
      let parts = words line
       in case filter ("any." `isPrefixOf`) parts of
            (p : _) ->
              let name = drop 4 p -- remove "any."
                  version = takeWhile (/= ',') $ drop 2 $ dropWhile (/= '=') line
               in Just $ name ++ "-" ++ filter (/= ' ') version
            [] -> Nothing

benchmark :: String -> Bool -> Bool -> IO ()
benchmark caseName profile useRg = do
  scriptDir <- (</> "etc") <$> getCurrentDirectory
  rootDir <- getCurrentDirectory
  let reposDir = scriptDir </> "repos"

  case Map.lookup caseName benchmarks of
    Nothing -> do
      putStrLn "invalid case; available cases:"
      mapM_ (\c -> putStrLn $ "  " ++ c) (Map.keys benchmarks)
      exitFailure
    Just scenario -> do
      let repoId = scenarioRepo scenario
          language = scenarioLang scenario
          symbol = scenarioSymbol scenario

      case Map.lookup repoId repos of
        Nothing -> do
          putStrLn $ "unknown repo: " ++ repoId
          exitFailure
        Just repoInfo -> do
          let gitUrl = repoGit repoInfo
              commit = repoCommit repoInfo
              repoPath = reposDir </> takeBaseName gitUrl

          putStrLn $ "scenario: " ++ caseName
          putStrLn $ "  git: " ++ gitUrl
          putStrLn $ "  commit: " ++ commit
          putStrLn $ "  language: " ++ language
          putStrLn $ "  symbol: " ++ symbol
          sep

          createDirectoryIfMissing True reposDir

          repoExists <- doesDirectoryExist repoPath
          unless repoExists $
            run $
              "git clone --filter=blob:none " ++ gitUrl ++ " " ++ repoPath

          run $ "git -C " ++ repoPath ++ " checkout --quiet " ++ commit

          setCurrentDirectory rootDir

          case (useRg, profile) of
            (True, _) -> do
              let cmd = "rg -t " ++ language ++ " --word-regexp " ++ symbol ++ " " ++ repoPath ++ " --column --no-heading"
              putStrLn "running rg..."
              run cmd
              sep
              putStrLn "running rg benchmark..."
              run $ "hyperfine -i --warmup 2 '" ++ cmd ++ "'"
            (_, True) -> do
              install True

              profExists <- doesFileExist "irk.prof"
              when profExists $ removeFile "irk.prof"

              sep
              putStrLn "profiling irk..."
              let cmd = "irk find -w " ++ repoPath ++ " -l " ++ language ++ " " ++ symbol ++ " +RTS -pj -N1 -RTS"

              (time, _) <- timeIt $ run cmd
              printf "elapsed time: %.3fs\n" time

              profExists' <- doesFileExist "irk.prof"
              when profExists' $ do
                now <- getCurrentTime
                let timestamp = formatTime defaultTimeLocale "%Y%m%d-%H%M%S" now
                    newname = "irk." ++ timestamp ++ ".prof"
                renameFile "irk.prof" newname
                sep
                putStrLn $ "profile saved to " ++ newname
            (_, _) -> do
              install False

              sep
              putStrLn "running irk..."
              let cmd = "irk find -w " ++ repoPath ++ " -l " ++ language ++ " " ++ symbol

              (time, _) <- timeIt $ run cmd
              printf "elapsed time: %.3fs\n" time

              sep
              putStrLn "running irk benchmark..."
              run $ "hyperfine --warmup 2 '" ++ cmd ++ "'"

timeIt :: IO a -> IO (Double, a)
timeIt io = do
  start <- getCurrentTime
  result <- io
  end <- getCurrentTime
  let diff = diffUTCTime end start
  return (realToFrac diff, result)

main :: IO ()
main = do
  cmd <- execParser opts
  case cmd of
    Install -> install False
    Vendor -> vendor
    Benchmark caseName profile useRg -> benchmark caseName profile useRg
